src/Model/Scope/AbstractScope.php

```diff
@@ -0,0 +1,194 @@
+<?php
+
+namespace atk4\data\Model\Scope;
+
+use atk4\core\InitializerTrait;
+use atk4\core\TrackableTrait;
+use atk4\data\Model;
+
+abstract class AbstractScope
+{
+    use InitializerTrait {
+        init as _init;
+    }
+    use TrackableTrait;
+
+    /**
+     * Defines if scope should be applied or not.
+     *
+     * @var bool
+     */
+    protected $active = true;
+
+    /**
+     * The model this scope belongs to.
+     *
+     * @var Model
+     */
+    public $model;
+
+    /**
+     * Contains the placeholder registry in $key => $options format.
+     *
+     * @var array
+     */
+    protected static $placeholders = [];
+
+    /**
+     * Register placeholoder for a value to be replaced
+     * The $options array may contain
+     * - label : string - the label to use when converting toWords
+     * - value : string|Callable - the actual value to be used when applying the scope
+     * If value is Callable the it is called with $model, $scope as arguments.
+     *
+     * @param string          $key
+     * @param string|callable $options
+     */
+    final public static function registerValuePlaceholder($key, $options)
+    {
+        self::$placeholders[$key] = is_array($options) ? $options : [
+            'label' => $key,
+            'value' => $options,
+        ];
+    }
+
+    /**
+     * Method is executed when the scope is added to Model using Model::add
+     * $this->owner in this case is the Model object.
+     */
+    public function init()
+    {
+        $this->_init();
+
+        $this->owner->scope()->addComponent($this);
+    }
+
+    /**
+     * Use a model.
+     *
+     * @param Model $model
+     *
+     * @return static
+     */
+    final public function on(Model $model)
+    {
+        return $this->setModel(clone $model);
+    }
+
+    /**
+     * Set the applicable model for the scope and its components.
+     *
+     * @param Model $model
+     *
+     * @return static
+     */
+    abstract public function setModel(Model $model = null);
+
+    /**
+     * Negate the scope object
+     * e.g from 'is' to 'is not'.
+     *
+     * @return $this
+     */
+    abstract public function negate();
```

Comment:
can this be documented in the main documentation?

Reply:
Done